File: client/ImageGenerator.php
<?php

class ImageGenerator {
    private $config;
    private $image;
    private $imgDir = __DIR__ . '/../static/images/';
    private $fontPath = __DIR__ . '/fonts/Roboto_Condensed-Bold.ttf';

    public function __construct($config) {
        $this->config = $config;
        $this->image = imagecreatetruecolor($config['canvas']['width'], $config['canvas']['height']);
        
        // Soporte para transparencia y mezcla de colores
        imagealphablending($this->image, true);
        imagesavealpha($this->image, true);
    }

    public function render() {
        // Primero renderizamos todas las imágenes (Capas inferiores)
        foreach ($this->config['elements'] as $el) {
            if ($el['type'] === 'image') $this->drawImage($el);
        }
        
        // Luego superponemos los textos (Capas superiores)
        foreach ($this->config['elements'] as $el) {
            if ($el['type'] === 'text') $this->drawText($el);
        }
    }

    private function drawImage($el) {
        $fullPath = $this->imgDir . $el['path'];
        if (!file_exists($fullPath)) return;

        $info = getimagesize($fullPath);
        $src = null;
        switch ($info[2]) {
            case IMAGETYPE_JPEG: $src = imagecreatefromjpeg($fullPath); break;
            case IMAGETYPE_PNG:  $src = imagecreatefrompng($fullPath);  break;
            case IMAGETYPE_WEBP: $src = imagecreatefromwebp($fullPath); break;
        }

        if ($src) {
            imagecopyresampled(
                $this->image, $src, 
                $el['x'], $el['y'], 0, 0, 
                $el['w'], $el['h'], $info[0], $info[1]
            );
            imagedestroy($src);
        }
    }

    private function drawText($el) {
        if (!file_exists($this->fontPath)) return;

        $data = $this->config['data'];
        $text = ($el['prefix'] ?? '') . ($data[$el['content_key']] ?? '');
        if (isset($el['suffix_key'])) $text .= " - " . $data[$el['suffix_key']];

        $colorArr = $this->hexToRgb($el['color']);
        $textColor = imagecolorallocate($this->image, $colorArr[0], $colorArr[1], $colorArr[2]);

        // Dibujar el borde (Outline) si está definido
        if (isset($el['outline_color']) && isset($el['outline_thickness'])) {
            $outArr = $this->hexToRgb($el['outline_color']);
            $outColor = imagecolorallocate($this->image, $outArr[0], $outArr[1], $outArr[2]);
            $this->drawOutline($el['size'], $el['x'], $el['y'], $outColor, $text, $el['outline_thickness']);
        }

        // Texto principal
        imagettftext($this->image, $el['size'], 0, $el['x'], $el['y'], $textColor, $this->fontPath, $text);
    }

    private function drawOutline($size, $x, $y, $color, $text, $thickness) {
        for ($dx = -$thickness; $dx <= $thickness; $dx++) {
            for ($dy = -$thickness; $dy <= $thickness; $dy++) {
                if ($dx == 0 && $dy == 0) continue;
                imagettftext($this->image, $size, 0, $x + $dx, $y + $dy, $color, $this->fontPath, $text);
            }
        }
    }

    private function hexToRgb($hex) {
        return sscanf($hex, "#%02x%02x%02x");
    }

    public function outputAvif($filename) {
        // Formato AVIF para máximo rendimiento técnico
        imageavif($this->image, $filename, 75);
        imagedestroy($this->image);
    }
}

File: frame_generator/image_generator.py
import os
from PIL import Image, ImageDraw, ImageFont

class ImageGenerator:
    def __init__(self, config):
        self.config = config
        base_path = os.path.dirname(os.path.abspath(__file__))
        # Ruta a las imágenes estáticas
        self.img_dir = os.path.abspath(os.path.join(base_path, '../static/images/'))
        # Ruta a la fuente (asegúrate de que el archivo existe en esta carpeta)
        self.font_path = os.path.join(base_path, 'fonts', 'Roboto_Condensed-Bold.ttf')
        
        canvas_cfg = config['canvas']
        self.image = Image.new('RGBA', (canvas_cfg['width'], canvas_cfg['height']), (0, 0, 0, 0))
        self.draw = ImageDraw.Draw(self.image)

    def render(self):
        # 1. Renderizar Imágenes (Capas fondo)
        for el in self.config['elements']:
            if el['type'] == 'image': self._draw_image(el)
        
        # 2. Renderizar Textos (Capas superiores)
        for el in self.config['elements']:
            if el['type'] == 'text': self._draw_text(el)

    def _draw_image(self, el):
        full_path = os.path.join(self.img_dir, el['path'])
        if not os.path.exists(full_path): return

        with Image.open(full_path) as img:
            img_resized = img.convert("RGBA").resize((el['w'], el['h']), Image.Resampling.LANCZOS)
            self.image.paste(img_resized, (el['x'], el['y']), img_resized)

    def _draw_text(self, el):
        if not os.path.exists(self.font_path):
            print(f"⚠️ Fuente no encontrada en: {self.font_path}")
            return

        data = self.config.get('data', {})
        
        # Lógica idéntica a PHP: prefix + contenido de la llave
        texto_base = data.get(el['content_key'], '')
        text = f"{el.get('prefix', '')}{texto_base}"
        
        # Soporte para suffix_key (como lo tenías en PHP)
        if 'suffix_key' in el and el['suffix_key'] in data:
            text += f" - {data[el['suffix_key']]}"

        font = ImageFont.truetype(self.font_path, el['size'])
        color = self._hex_to_rgb(el['color'])

        # Dibujar Outline si existe
        if 'outline_color' in el and 'outline_thickness' in el:
            out_color = self._hex_to_rgb(el['outline_color'])
            self.draw.text(
                (el['x'], el['y']), text, font=font, fill=color,
                stroke_width=el['outline_thickness'], stroke_fill=out_color
            )
        else:
            self.draw.text((el['x'], el['y']), text, font=font, fill=color)

    def _hex_to_rgb(self, hex_str):
        hex_str = hex_str.lstrip('#')
        return tuple(int(hex_str[i:i+2], 16) for i in (0, 2, 4))

    def output_png(self, filename):
        os.makedirs(os.path.dirname(filename), exist_ok=True)
        self.image.save(filename, "PNG")

    def output_avif(self, filename):
        os.makedirs(os.path.dirname(filename), exist_ok=True)
        try:
            self.image.convert("RGB").save(filename, "AVIF", quality=75)
        except:
            self.image.save(filename.replace('.avif', '.webp'), "WEBP", quality=75)
